dm/dm/worker/source_worker.go

```diff
@@ -798,35 +801,43 @@ func (w *SourceWorker) handleSubTaskStage(ctx context.Context, stageCh chan ha.S
 // operateSubTaskStage returns TaskOp.String() additionally to record metrics.
 func (w *SourceWorker) operateSubTaskStage(stage ha.Stage, subTaskCfg config.SubTaskConfig) (string, error) {
 	var op pb.TaskOp
-	switch {
-	case stage.Expect == pb.Stage_Running, stage.Expect == pb.Stage_Paused:
-		if st := w.subTaskHolder.findSubTask(stage.Task); st == nil {
-			// create the subtask for expected running and paused stage.
-			log.L().Info("start to create subtask", zap.String("sourceID", subTaskCfg.SourceID), zap.String("task", subTaskCfg.Name))
-
+	log.L().Info("operateSubTaskStage",
+		zap.String("sourceID", subTaskCfg.SourceID),
+		zap.String("task", subTaskCfg.Name),
+		zap.Stringer("stage", stage))
+
+	// for new added subtask
+	if st := w.subTaskHolder.findSubTask(stage.Task); st == nil {
+		switch stage.Expect {
+		case pb.Stage_Running, pb.Stage_Paused, pb.Stage_Stopped:
+			log.L().Info("start to create subtask in operateSubTaskStage", zap.String("sourceID", subTaskCfg.SourceID), zap.String("task", subTaskCfg.Name))
 			expectValidatorStage, err := getExpectValidatorStage(subTaskCfg.ValidatorCfg, w.etcdClient, stage.Source, stage.Task, stage.Revision)
 			if err != nil {
 				return opErrTypeBeforeOp, terror.Annotate(err, "fail to get validator stage from etcd")
 			}
-
-			err = w.StartSubTask(&subTaskCfg, stage.Expect, expectValidatorStage, true)
-			return opErrTypeBeforeOp, err
-		}
-		if stage.Expect == pb.Stage_Running {
-			op = pb.TaskOp_Resume
-		} else if stage.Expect == pb.Stage_Paused {
-			op = pb.TaskOp_Pause
+			return opErrTypeBeforeOp, w.StartSubTask(&subTaskCfg, stage.Expect, expectValidatorStage, true)
+		default:
```

Comment:
i left a comment [here](https://github.com/Ehco1996/ticdc/blob/c608c0b6c423e5e08da4601c5c9c1426cc461748/dm/dm/worker/source_worker.go#L820), this is err from operateSubTask is expected, call should not operate a  task that not in `subTaskHolder`

Reply:
Oh I see the old logic only handles Stage_Running and Stage_Paused. But I think the logic of deciding if the expected stage is valid should be put inside `StartSubTask` and `OperateSubTask`, not in `operateSubTaskStage`. In other words, if the subtask exists, we simply forward expecte stage to `OperateSubTask`, otherwise to `StartSubTask`.

It's OK we change it in future.